File: MapsMeasure/src/de/j4velin/mapsmeasure/Map.java
/*
 * Copyright 2014 Thomas Hoffmann
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.j4velin.mapsmeasure;

import java.io.IOException;
import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.Stack;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.GoogleMap.OnMapClickListener;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.CameraPosition;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Polygon;
import com.google.android.gms.maps.model.PolygonOptions;
import com.google.android.gms.maps.model.Polyline;
import com.google.android.gms.maps.model.PolylineOptions;
import com.google.maps.android.SphericalUtil;

import android.content.Context;
import android.graphics.Color;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.location.LocationManager;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.FragmentActivity;
import android.view.KeyEvent;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.widget.TextView.OnEditorActionListener;
import android.widget.Toast;

public class Map extends FragmentActivity {

	// the map to draw to
	private GoogleMap mMap;

	// the stacks - everytime the user touches the map, an entry is pushed
	private Stack<LatLng> trace = new Stack<LatLng>();
	private Stack<Polyline> lines = new Stack<Polyline>();
	private Stack<Polygon> points = new Stack<Polygon>();

	private float distance; // in meters
	private TextView distanceTv; // the view displaying the distance & unit

	private final static int COLOR_LINE = Color.argb(128, 0, 0, 0), COLOR_POINT = Color.argb(128, 255, 0, 0);

	private final static NumberFormat formatter = NumberFormat.getInstance(Locale.getDefault());

	private String unit = "m";

	/**
	 * Update the distance textview.
	 * 
	 * Should be called when the distance or the unit changes
	 */
	private void updateDistance() {
		float d = distance;
		if (unit.equals("m")) {
			if (d > 1000) {
				unit = "km";
				updateDistance();
				return;
			}
		} else if (unit.equals("km")) {
			if (d < 1000) {
				unit = "m";
				updateDistance();
				return;
			}
			d /= 1000;
		} else if (unit.equals("yd")) {
			if (distance > 1609) {
				unit = "mi";
				updateDistance();
				return;
			}
			d = distance / 0.9144f;
		} else if (unit.equals("mi")) {
			if (d < 1609) {
				unit = "yd";
				updateDistance();
				return;
			}
			d = distance / 1609.344f;
		}

		distanceTv.setText(formatter.format(Math.max(0, d)) + " " + unit);
	}

	@Override
	protected void onRestoreInstanceState(final Bundle savedInstanceState) {
		super.onRestoreInstanceState(savedInstanceState);
		unit = savedInstanceState.getString("unit");
		@SuppressWarnings("unchecked")
		Stack<LatLng> tmp = (Stack<LatLng>) savedInstanceState.getSerializable("trace");
		Iterator<LatLng> it = tmp.iterator();
		while (it.hasNext()) {
			addPoint(it.next());
		}
		mMap.moveCamera(CameraUpdateFactory.newCameraPosition((CameraPosition) savedInstanceState.getParcelable("position")));
	}

	@Override
	protected void onSaveInstanceState(final Bundle outState) {
		outState.putSerializable("trace", trace);
		outState.putString("unit", unit);
		outState.putParcelable("position", mMap.getCameraPosition());
		super.onSaveInstanceState(outState);
	}

	/**
	 * Adds a new point, calculates the new distance and draws the point and a
	 * line to it
	 * 
	 * @param p
	 *            the new point
	 */
	private void addPoint(final LatLng p) {
		if (!trace.isEmpty()) {
			lines.push(mMap.addPolyline(new PolylineOptions().color(COLOR_LINE).add(trace.peek()).add(p)));
			distance += SphericalUtil.computeDistanceBetween(p, trace.peek());
			updateDistance();
		}
		points.push(drawCircle(p));
		trace.push(p);
	}

	/**
	 * Removes the last added point, the line to it and updates the distance
	 */
	private void removeLast() {
		if (trace.isEmpty())
			return;
		points.pop().remove();
		LatLng remove = trace.pop();
		if (!trace.isEmpty())
			distance -= SphericalUtil.computeDistanceBetween(remove, trace.peek());
		if (!lines.isEmpty())
			lines.pop().remove();
		updateDistance();
	}

	@Override
	public void onCreate(final Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_map);

		mMap = ((SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.map)).getMap();
		distanceTv = (TextView) findViewById(R.id.distance);
		distanceTv.setText("0 " + unit);
		distanceTv.setOnClickListener(new OnClickListener() {
			@Override
			public void onClick(final View v) {
				if (unit.equals("m") || unit.equals("km")) {
					unit = "yd";
				} else {
					unit = "m";
				}
				// will change to km/mi if needed
				updateDistance();
			}
		});

		final EditText search = (EditText) findViewById(R.id.search);
		search.setOnEditorActionListener(new OnEditorActionListener() {
			@Override
			public boolean onEditorAction(final TextView v, int actionId, KeyEvent event) {
				new GeocoderTask().execute(search.getText().toString());
				InputMethodManager inputManager = (InputMethodManager) Map.this.getSystemService(Context.INPUT_METHOD_SERVICE);
				inputManager.hideSoftInputFromWindow(Map.this.getCurrentFocus().getWindowToken(),
						InputMethodManager.HIDE_NOT_ALWAYS);
				return true;
			}
		});

		findViewById(R.id.delete).setOnClickListener(new OnClickListener() {
			@Override
			public void onClick(final View v) {
				removeLast();
			}
		});

		formatter.setMaximumFractionDigits(2);

		mMap.setOnMapClickListener(new OnMapClickListener() {
			@Override
			public void onMapClick(final LatLng center) {
				addPoint(center);
			}
		});

		// KitKat translucent decor enabled? -> Add some margin/padding to the
		// textview and the map
		if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT) {
			View topleft = findViewById(R.id.topleft);
			FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) topleft.getLayoutParams();
			lp.setMargins(10, Util.getStatusBarHeight(this) + 10, 0, 0);
			topleft.setLayoutParams(lp);

			mMap.setPadding(0, Util.getStatusBarHeight(this), 0, Util.getNavigationBarHeight(this));
		}

		mMap.setMyLocationEnabled(true);

		LatLng userLocation = null; // location to move to

		if (getIntent().getDoubleExtra("lon", 1000) < 999) {
			double lon = getIntent().getDoubleExtra("lon", 0);
			double lat = getIntent().getDoubleExtra("lat", 0);
			userLocation = new LatLng(lat, lon);
		} else {
			Location location = mMap.getMyLocation();
			if (location == null) {
				LocationManager locationManager = (LocationManager) getSystemService(LOCATION_SERVICE);
				List<String> matchingProviders = locationManager.getAllProviders();
				float accuracy, bestAccuracy = Float.MAX_VALUE;
				long time, minTime = 0;
				Location tmp;
				for (String provider : matchingProviders) {
					tmp = locationManager.getLastKnownLocation(provider);
					if (tmp != null) {
						accuracy = tmp.getAccuracy();
						time = tmp.getTime();

						if ((time > minTime && accuracy < bestAccuracy)) {
							location = tmp;
							bestAccuracy = accuracy;
							minTime = time;
						} else if (time < minTime && bestAccuracy == Float.MAX_VALUE && time > minTime) {
							location = tmp;
							minTime = time;
						}
					}
				}
			}
			if (location != null) {
				userLocation = new LatLng(location.getLatitude(), location.getLongitude());
			}
		}

		if (userLocation != null) {
			mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(userLocation, 16));
		}
	}

	private LatLng getPoint(final LatLng center, final float radius, final double angle) {
		double east = radius * Math.cos(angle);
		double north = radius * Math.sin(angle);

		double cLat = center.latitude;
		double cLng = center.longitude;
		double latRadius = SphericalUtil.EARTH_RADIUS * Math.cos(cLat / 180 * Math.PI);

		double newLat = cLat + (north / SphericalUtil.EARTH_RADIUS / Math.PI * 180);
		double newLng = cLng + (east / latRadius / Math.PI * 180);

		return new LatLng(newLat, newLng);
	}

	/**
	 * Draws a circle at the given point.
	 * 
	 * Should be called when the users touches the map and adds an entry to the
	 * stacks
	 * 
	 * @param center
	 *            the point where the user clicked
	 * @return the drawn Polygon
	 */
	private Polygon drawCircle(final LatLng center) {
		int totalPonts = 30; // number of corners of the pseudo-circle
		List<LatLng> points = new ArrayList<LatLng>(totalPonts);
		for (int i = 0; i < totalPonts; i++) {
			points.add(getPoint(center, 10, i * 2 * Math.PI / totalPonts));
		}
		return mMap.addPolygon(new PolygonOptions().addAll(points).strokeWidth(0).fillColor(COLOR_POINT));
	}

	/**
	 * 
	 * Based on
	 * http://wptrafficanalyzer.in/blog/android-geocoding-showing-user-input
	 * -location-on-google-map-android-api-v2/
	 * 
	 * @author George Mathew
	 * 
	 */
	private class GeocoderTask extends AsyncTask<String, Void, Address> {

		@Override
		protected Address doInBackground(final String... locationName) {
			// Creating an instance of Geocoder class
			Geocoder geocoder = new Geocoder(getBaseContext());
			try {
				// Get only the best result that matches the input text
				List<Address> addresses = geocoder.getFromLocationName(locationName[0], 1);
				return addresses != null && !addresses.isEmpty() ? addresses.get(0) : null;
			} catch (IOException e) {
				Toast.makeText(Map.this, "Error: " + e.getMessage(), Toast.LENGTH_SHORT).show();
				e.printStackTrace();
				return null;
			}
		}

		@Override
		protected void onPostExecute(final Address address) {
			if (address == null) {
				Toast.makeText(getBaseContext(), "No Location found", Toast.LENGTH_SHORT).show();
			} else {
				mMap.animateCamera(CameraUpdateFactory.newLatLng(new LatLng(address.getLatitude(), address.getLongitude())));
			}
		}
	}
}

File: src/main/java/de/j4velin/mapsmeasure/Map.kt
/*
 * Copyright 2014 Thomas Hoffmann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package de.j4velin.mapsmeasure

import android.Manifest
import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.ActivityNotFoundException
import android.content.Intent
import android.graphics.Color
import android.net.Uri
import android.os.BadParcelableException
import android.os.Bundle
import android.util.Log
import android.util.TypedValue
import android.view.Menu
import android.view.View
import android.widget.FrameLayout
import android.widget.ListView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.content.PermissionChecker
import androidx.core.net.toUri
import androidx.core.view.GravityCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.BitmapDescriptor
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.gms.maps.model.Polygon
import com.google.android.gms.maps.model.PolygonOptions
import com.google.android.gms.maps.model.Polyline
import com.google.android.gms.maps.model.PolylineOptions
import kotlinx.coroutines.launch
import kotlin.math.max
import kotlin.math.min

/**
 * The measure screen. All state lives in the [MeasureViewModel]; this activity draws it and
 * forwards the user's input.
 */
class Map : FragmentActivity(), OnMapReadyCallback {

    private lateinit var viewModel: MeasureViewModel

    private var drawerLayout: DrawerLayout? = null
    private lateinit var valueTv: TextView // the view displaying the distance/area & unit
    private lateinit var drawerListAdapter: DrawerListAdapter
    private var renderedType: MeasureType? = null
    private var renderedMapType: Int? = null

    // what is currently drawn on the map
    private var drawnTrace: List<LatLng> = emptyList()
    private val lines = ArrayList<Polyline>() // lines[i] connects drawnTrace[i] and drawnTrace[i + 1]
    private val points = ArrayList<Marker?>() // points[i] marks drawnTrace[i]
    private var areaOverlay: Polygon? = null
    private var drawnArea: List<LatLng>? = null
    private var marker: BitmapDescriptor? = null

    private var navBarOnRight = false
    private var drawerSize = 0
    private var statusbar = 0
    private var navBarHeight = 0

    private var googleMap: GoogleMap? = null

    // what to do once we have the location permission
    private var pendingLocationAction: (() -> Unit)? = null

    private val locationPermissionRequest =
        registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) {
            onLocationPermissionResult()
        }

    fun closeDrawer() {
        drawerLayout?.closeDrawers()
    }

    /**
     * Replaces the current points with the ones from the given file
     */
    fun loadTrace(uri: Uri) = viewModel.loadTrace(uri)

    /**
     * Searches for the given place and moves the map there
     */
    fun searchLocation(query: String) = viewModel.search(query)

    fun setMetric(metric: Boolean) = viewModel.setMetric(metric)

    override fun onCreate(savedInstanceState: Bundle?) {
        try {
            super.onCreate(savedInstanceState)
        } catch (bpe: BadParcelableException) {
            if (BuildConfig.DEBUG) Log.d(LOG_TAG, "can not restore state", bpe)
        }
        viewModel = ViewModelProvider(this, MeasureViewModel.Factory)[MeasureViewModel::class.java]
        init()
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.uiState.collect { render(it) }
            }
        }
    }

    /**
     * Initializes everything
     */
    private fun init() {
        setContentView(R.layout.activity_map)

        val topCenterOverlay = findViewById<View>(R.id.topCenterOverlay)
        val drawerLayout: DrawerLayout? = findViewById(R.id.drawer_layout)
        this.drawerLayout = drawerLayout

        val menuButton: View? = findViewById(R.id.menu)
        menuButton?.setOnClickListener { drawerLayout?.openDrawer(GravityCompat.START) }

        drawerLayout?.let {
            it.setDrawerShadow(R.drawable.drawer_shadow, GravityCompat.START)

            it.addDrawerListener(object : DrawerLayout.DrawerListener {
                private var menuButtonVisible = true

                override fun onDrawerStateChanged(newState: Int) {}

                override fun onDrawerSlide(drawerView: View, slideOffset: Float) {
                    topCenterOverlay.alpha = 1 - slideOffset
                    if (menuButtonVisible && menuButton != null && slideOffset > 0) {
                        menuButton.visibility = View.INVISIBLE
                        menuButtonVisible = false
                    }
                }

                override fun onDrawerOpened(drawerView: View) {
                    topCenterOverlay.visibility = View.INVISIBLE
                }

                override fun onDrawerClosed(drawerView: View) {
                    topCenterOverlay.visibility = View.VISIBLE
                    if (menuButton != null) {
                        menuButton.visibility = View.VISIBLE
                        menuButtonVisible = true
                    }
                }
            })
        }

        (supportFragmentManager.findFragmentById(R.id.map) as SupportMapFragment).getMapAsync(this)

        valueTv = findViewById(R.id.distance)
        valueTv.setOnClickListener {
            viewModel.toggleType()
            closeDrawer()
        }

        val delete = findViewById<View>(R.id.delete)
        delete.setOnClickListener { viewModel.removeLastPoint() }
        delete.setOnLongClickListener {
            val count = viewModel.uiState.value.trace.size
            AlertDialog.Builder(this)
                .setMessage(resources.getQuantityString(R.plurals.delete_all, count, count))
                .setPositiveButton(android.R.string.yes) { dialog, _ ->
                    viewModel.clear()
                    dialog.dismiss()
                }
                .setNegativeButton(android.R.string.no) { dialog, _ -> dialog.dismiss() }
                .create().show()
            true
        }

        // Drawer stuff
        val drawerList = findViewById<ListView>(R.id.left_drawer)
        drawerListAdapter = DrawerListAdapter(this)
        drawerList.adapter = drawerListAdapter
        drawerList.divider = null
        drawerList.setOnItemClickListener { _, _, position, _ ->
            val state = viewModel.uiState.value
            when (position) {
                2 -> { // Units
                    Dialogs.getUnits(this, state.metric, state.distance, state.area).show()
                    closeDrawer()
                }
                3 -> {
                    viewModel.setType(MeasureType.DISTANCE)
                    closeDrawer()
                }
                4 -> {
                    viewModel.setType(MeasureType.AREA)
                    closeDrawer()
                }
                6, 7, 8 -> {
                    viewModel.setMapType(MAP_TYPE_AT_POSITION.getValue(position))
                    closeDrawer()
                }
                10 -> { // save
                    Dialogs.getSaveNShare(this, state.trace).show()
                    closeDrawer()
                }
                11 -> { // more apps
                    try {
                        startActivity(
                            Intent(Intent.ACTION_VIEW, "market://search?q=pub:j4velin".toUri())
                                .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        )
                    } catch (anf: ActivityNotFoundException) {
                        startActivity(
                            Intent(
                                Intent.ACTION_VIEW,
                                "https://play.google.com/store/apps/developer?id=j4velin".toUri()
                            ).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                        )
                    }
                }
                12 -> { // about
                    Dialogs.getAbout(this).show()
                    closeDrawer()
                }
            }
        }

        drawerSize = if (drawerLayout == null) {
            TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 200f, resources.displayMetrics)
                .toInt()
        } else {
            0
        }

        // the window is drawn behind the translucent system bars -> move the overlays out of their way
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(android.R.id.content)) { _, windowInsets ->
            val bars = windowInsets.getInsets(WindowInsetsCompat.Type.systemBars())
            statusbar = bars.top
            // in landscape on phones, the navigation bar might be at the
            // right side, reducing the available display width
            navBarOnRight = bars.right > 0
            navBarHeight = if (navBarOnRight) bars.right else bars.bottom

            val lp = topCenterOverlay.layoutParams as FrameLayout.LayoutParams
            lp.setMargins(0, statusbar + 10, 0, 0)
            topCenterOverlay.layoutParams = lp

            drawerList.setPadding(0, statusbar + 10, 0, 0)
            if (navBarOnRight) {
                drawerListAdapter.setMarginBottom(0)
                menuButton?.setPadding(0, 0, 0, 0)
            } else {
                drawerListAdapter.setMarginBottom(navBarHeight)
                menuButton?.setPadding(0, 0, 0, navBarHeight)
            }

            updateMapPadding()
            windowInsets
        }
    }

    /**
     * Shows the state in the views outside of the map
     */
    private fun render(state: MeasureUiState) {
        valueTv.text = state.formattedValue()
        if (state.type != renderedType) {
            drawerListAdapter.changeType(state.type)
            renderedType = state.type
        }
        if (state.mapType != renderedMapType) {
            drawerListAdapter.changeView(state.mapType)
            renderedMapType = state.mapType
        }
    }

    @SuppressLint("MissingPermission")
    override fun onMapReady(map: GoogleMap) {
        googleMap = map
        marker = BitmapDescriptorFactory.fromResource(R.drawable.marker)

        // back where it was before the activity got recreated
        viewModel.camera?.let { map.moveCamera(CameraUpdateFactory.newCameraPosition(it)) }

        map.setOnMarkerClickListener {
            viewModel.addPoint(it.position)
            true
        }
        map.setOnMapClickListener { viewModel.addPoint(it) }
        map.setOnCameraIdleListener { viewModel.onCameraIdle(map.cameraPosition) }

        map.uiSettings.isMyLocationButtonEnabled = true
        map.setOnMyLocationButtonClickListener {
            withLocationPermission { viewModel.onMyLocationButton() }
            true
        }
        if (hasLocationPermission()) {
            map.isMyLocationEnabled = true
        }

        updateMapPadding()

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch { viewModel.uiState.collect { draw(map, it) } }
                viewModel.events.collect { handle(map, it) }
            }
        }

        if (viewModel.consumeFirstStart()) {
            val data = intent.data
            if (Intent.ACTION_VIEW == intent.action && data != null) {
                // opened with a csv file
                viewModel.loadTrace(data)
            } else {
                withLocationPermission { viewModel.centerOnCurrentLocation() }
            }
        }
    }

    /**
     * Shows the state on the map. Only the points and lines which changed are redrawn.
     */
    private fun draw(map: GoogleMap, state: MeasureUiState) {
        if (map.mapType != state.mapType) map.mapType = state.mapType

        val trace = state.trace
        if (trace != drawnTrace) {
            // the points both traces start with stay
            var keep = 0
            val common = min(trace.size, drawnTrace.size)
            while (keep < common && trace[keep] == drawnTrace[keep]) keep++

            while (points.size > keep) points.removeAt(points.lastIndex)?.remove()
            while (lines.size > max(keep - 1, 0)) lines.removeAt(lines.lastIndex).remove()

            for (i in keep until trace.size) {
                if (i > 0) {
                    lines.add(
                        map.addPolyline(
                            PolylineOptions().color(COLOR_LINE).width(LINE_WIDTH)
                                .add(trace[i - 1]).add(trace[i])
                        )
                    )
                }
                points.add(
                    map.addMarker(
                        MarkerOptions().position(trace[i]).flat(true).anchor(0.5f, 0.5f).icon(marker)
                    )
                )
            }
            drawnTrace = trace
        }

        val area = if (state.type == MeasureType.AREA && trace.size >= 3) trace else null
        if (area != drawnArea) {
            areaOverlay?.remove()
            areaOverlay = area?.let {
                map.addPolygon(PolygonOptions().addAll(it).strokeWidth(0f).fillColor(COLOR_POINT))
            }
            drawnArea = area
        }
    }

    private fun handle(map: GoogleMap, event: MeasureEvent) {
        when (event) {
            is MeasureEvent.MoveCamera -> {
                val update = CameraUpdateFactory.newLatLngZoom(event.target, event.zoom)
                if (event.animate) map.animateCamera(update) else map.moveCamera(update)
            }
            is MeasureEvent.Message ->
                Toast.makeText(this, event.text, Toast.LENGTH_SHORT).show()
            is MeasureEvent.Error ->
                Toast.makeText(
                    this,
                    getString(
                        R.string.error,
                        event.exception.javaClass.simpleName + "\n" + event.exception.message
                    ),
                    Toast.LENGTH_LONG
                ).show()
        }
    }

    /**
     * Keeps the map controls out from under the system bars
     */
    private fun updateMapPadding() {
        if (navBarOnRight) {
            // in landscape on phones, the navigation bar might be at the
            // right side, reducing the available display width
            googleMap?.setPadding(drawerSize, statusbar, navBarHeight, 0)
        } else {
            googleMap?.setPadding(0, statusbar, 0, navBarHeight)
        }
    }

    /**
     * Runs the given action now if we have the location permission, or asks for it first
     */
    private fun withLocationPermission(action: () -> Unit) {
        if (hasLocationPermission()) {
            action()
        } else {
            pendingLocationAction = action
            locationPermissionRequest.launch(
                arrayOf(
                    Manifest.permission.ACCESS_COARSE_LOCATION,
                    Manifest.permission.ACCESS_FINE_LOCATION
                )
            )
        }
    }

    @SuppressLint("MissingPermission")
    private fun onLocationPermissionResult() {
        if (hasLocationPermission()) {
            pendingLocationAction?.invoke()
            googleMap?.isMyLocationEnabled = true
        } else {
            viewModel.moveToLastPosition()
        }
        pendingLocationAction = null
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val drawerLayout = drawerLayout ?: return true
        if (drawerLayout.isDrawerOpen(GravityCompat.START)) drawerLayout.closeDrawers()
        else drawerLayout.openDrawer(GravityCompat.START)
        return false
    }

    /**
     * @return true, if the user granted at least approximate location access, which is
     * enough for the location button and to center the map
     */
    private fun hasLocationPermission(): Boolean =
        PermissionChecker.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) ==
                PermissionChecker.PERMISSION_GRANTED ||
                PermissionChecker.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) ==
                PermissionChecker.PERMISSION_GRANTED

    companion object {
        const val LOG_TAG = "MapsMeasure"

        private val COLOR_LINE = Color.argb(128, 0, 0, 0)
        private val COLOR_POINT = Color.argb(128, 255, 0, 0)
        private const val LINE_WIDTH = 5f

        private val MAP_TYPE_AT_POSITION = mapOf(
            6 to GoogleMap.MAP_TYPE_NORMAL,
            7 to GoogleMap.MAP_TYPE_HYBRID,
            8 to GoogleMap.MAP_TYPE_TERRAIN
        )
    }
}
